File: game-logic-core/src/main/scala/fwc/game/GameState.scala
package fwc.game

import fwc.{JsonParsable, JsonSerializable}
import fwc.game.actionPhase.{Combat, DiscardedHouseCards, DominanceTokensUsage}
import fwc.game.eventsPhase.cards.BoardCards
import fwc.game.board.{Armies, DominanceTokenType, TileNumber, Tracks}
import fwc.game.eventsPhase.{Bids, PowerTokens, Supplies, UsedMusteringPoints}
import fwc.game.houses.HouseType
import fwc.game.phases.{PhasePlanning, SubPhase}
import fwc.game.planningPhase.{AvailableOrders, Order, OrderConsolidatePower, OrderType, PlacedOrders}
import fwc.gameLoading.BoardTile
import ujson.Value

import scala.collection.mutable
import scala.util.Try

case class GameState(
                      subPhase: SubPhase,
                      armies: Armies,
                      placedOrders: PlacedOrders,
                      tracks: Tracks,
                      supplies: Supplies,
                      discardedHouseCards: DiscardedHouseCards,
                      powerTokens: PowerTokens,
                    //TODO need to add used board cards
                      boardCards: BoardCards,
                      dominanceTokensUsage: DominanceTokensUsage,
                      usedMusteringPoints: UsedMusteringPoints,
                      availableOrders: AvailableOrders,
                      bids: Bids,
                      combat: Combat = null,
                      wildlingCounter: Int = 6,
                      wildlingsStartedFrom12Points: Option[Boolean] = None,
                      roundCounter: Int = 1,
                      winner: Option[HouseType] = None
                    ) extends JsonSerializable {

  def toPersonalJson(myHouse: HouseType): ujson.Value = {
    val placedOrdersP = 
      if subPhase.getMainPhase == PhasePlanning then
        placedOrders.copy(placedOrders = placedOrders.placedOrders.map(
        (houseType: HouseType, orders: Map[TileNumber, Order]) =>
          if houseType == myHouse then
            houseType -> orders
          else
            houseType -> orders.map(
              (tileNum: TileNumber, order: Order) =>
                tileNum -> Order(OrderConsolidatePower)
            )
        ))
      else placedOrders
    ujson.Obj(
      toCleanJson.value ++ mutable.LinkedHashMap[String, ujson.Value](
        "placedOrders" -> placedOrdersP.toJson,
      )
    )
  }

  def toJson: ujson.Value = {
    ujson.Obj(
      toCleanJson.value ++ mutable.LinkedHashMap[String, ujson.Value](
        "boardCards" -> boardCards.toJson,
        "placedOrders" -> placedOrders.toJson,
        "availableOrders" -> availableOrders.toJson,
        "bids" -> bids.toJson,
      )
    )
  }

  def toCleanJson: ujson.Obj = {
    ujson.Obj(
      "subPhase" -> subPhase.toJson,
      "armies" -> armies.toJson,
      "tracks" -> tracks.toJson,
      "supplies" -> supplies.toJson,
      "discardedHouseCards" -> discardedHouseCards.toJson,
      "powerTokens" -> powerTokens.toJson,
      "dominanceTokensUsage" -> dominanceTokensUsage.toJson,
      "usedMusteringPoints" -> usedMusteringPoints.toJson,
      "combat" -> (if combat == null then ujson.Null else combat.toJson),
      "wildlingCounter" -> wildlingCounter,
      "wildlingsStartedFrom12Points" ->
        (if wildlingsStartedFrom12Points.isEmpty then ujson.Null else wildlingsStartedFrom12Points.head),
      "roundCounter" -> roundCounter,
    )
  }

}

object GameState extends JsonParsable {
  override def fromJson(json: Value): GameState = {
    GameState(
      SubPhase.fromJson(json.obj("subPhase")),
      Armies.fromJson(json.obj("armies")),
      PlacedOrders.fromJson(json.obj("placedOrders")),
      Tracks.fromJson(json.obj("tracks")),
      Supplies.fromJson(json.obj("supplies")),
      DiscardedHouseCards.fromJson(json.obj("discardedHouseCards")),
      PowerTokens.fromJson(json.obj("powerTokens")),
      BoardCards.fromJson(json.obj("boardCards")),
      DominanceTokensUsage.fromJson(json.obj("dominanceTokensUsage")),
      UsedMusteringPoints.fromJson(json.obj("usedMusteringPoints")),
      AvailableOrders.fromJson(json.obj("availableOrders")),
      Bids.fromJson(json.obj("bids")),
      Combat.fromJson(json("combat")),
      json.obj("wildlingCounter").num.toInt,
      Try[Option[Boolean]](json("wildlingsStartedFrom12Points").boolOpt).getOrElse(None),
      json.obj("roundCounter").num.toInt
    )

  }
}

File: bots/server/game_state/game_state.py
from server.game_state.armies import Armies
from server.game_state.placed_orders import PlacedOrders
from server.game_state.tracks import Tracks
from server.game_state.supplies import Supplies
from server.game_state.discarded_house_cards import DiscardedHouseCards
from server.game_state.power_tokens import PowerTokens
from server.game_state.used_mustering_points import UsedMusteringPoints
from server.game_state.available_orders import AvailableOrders
from server.game_state.combat import Combat


class GameState:
    def __init__(
            self,
            armies: Armies,
            placed_orders: PlacedOrders,
            tracks: Tracks,
            supplies: Supplies,
            discarded_house_cards: DiscardedHouseCards,
            power_tokens: PowerTokens,
            used_mustering_points: UsedMusteringPoints,
            available_orders: AvailableOrders,
            combat: Combat,
            wildling_counter: int = 6,
            round_counter: int = 1
    ):
        self.armies = armies 
        self.placed_orders = placed_orders 
        self.tracks = tracks 
        self.supplies = supplies 
        self.discarded_house_cards = discarded_house_cards 
        self.power_tokens = power_tokens
        self.used_mustering_points = used_mustering_points 
        self.available_orders = available_orders
        self.combat = combat 
        self.wildling_counter = wildling_counter
        self.round_counter = round_counter

    @classmethod
    def from_json(cls, json):
        return cls(
            Armies(**json['armies']),
            PlacedOrders(**json['placedOrders']) if 'placedOrders' in json else PlacedOrders(),
            Tracks(**json['tracks']),
            Supplies(**json['supplies']),
            DiscardedHouseCards(**json['discardedHouseCards']),
            PowerTokens(**json['powerTokens']),
            UsedMusteringPoints(**json['usedMusteringPoints']),
            AvailableOrders(**json['availableOrders']) if 'availableOrders' in json else AvailableOrders(),
            Combat.from_json(json['combat']) if ('combat' in json and json['combat'] is not None) else None,
            json['wildlingCounter'],
            json['roundCounter'],
        )